File: zb1/codingTest/220317/(2)SecondLargestBinaryNumber/SecondLargestBinaryNumber.py
def solution(n):
    """
    :param n: int
    :return: int
    """

    result = 0
    for i in range(1, n + 1):
        result += getSecondLargestBitNumber(i)
    return result


def getSecondLargestBitNumber(n):
    pivot = 1 << 15
    count = 0

    while pivot > 0:
        if pivot & n:
            count += 1
            if count == 2: return pivot

        pivot >>= 1

    return 0

File: zb1/codingTest/220317/(2)SecondLargestBinaryNumber/SecondLargestBinaryNumber.js
/**
 * @param n {number}
 * @returns {number}
 */
function solution(n) {
  let result = 0
  for (let i = 1; i <= n; i++) {
    result += getSecondLargestBitNumber(i)
  }
  return result
}

function getSecondLargestBitNumber(n) {
  let max = 1 << 30
  let count = 0

  for (; max > 0; max >>= 1) {
    if (max & n) {
      if (count === 1) {
        count++
        continue
      }

      count++
      n -= max
    }
  }

  return n
}

solution